round_robin/round_robin.go

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 CloudWeGo Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package roundrobin
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/cloudwego/hertz/pkg/app/client/discovery"
+	"github.com/cloudwego/hertz/pkg/app/client/loadbalance"
+	"golang.org/x/sync/singleflight"
+)
+
+type roundRobinBalancer struct {
+	cachedInfo sync.Map
+	sfg        singleflight.Group
+}
+
+type roundRobinInfo struct {
+	instances []discovery.Instance
+	index     atomic.Value
+}
+
+// NewRoundRobinBalancer creates a loadbalancer using round-robin algorithm.
+func NewRoundRobinBalancer() loadbalance.Loadbalancer {
+	lb := &roundRobinBalancer{}
+	return lb
+}
+
+// Pick implements the Loadbalancer interface.
+func (rr *roundRobinBalancer) Pick(e discovery.Result) discovery.Instance {
+	ri, ok := rr.cachedInfo.Load(e.CacheKey)
+	if !ok {
+		ri, _, _ = rr.sfg.Do(e.CacheKey, func() (interface{}, error) {
+			aV := atomic.Value{}
+			aV.Store(0)
+			return &roundRobinInfo{
+				instances: e.Instances,
+				index:     aV,
+			}, nil
+		})
+		rr.cachedInfo.Store(e.CacheKey, ri)
+	}
+
+	r := ri.(*roundRobinInfo)
+	if len(r.instances) == 0 {
+		return nil
+	}
+
+	lens := len(r.instances)
+	if r.index.Load().(int) >= lens {
+		r.index.Store(0)
+	}
+
+	instance := r.instances[r.index.Load().(int)]
```

Comment:
It won't take effect. You can use atomic.AddUint32 first and use the return value -1 as the index

Reply:
And there's no need to zero it out. Use uint instead and let it increase all the time.

Reply:
atomic.AddUint32 will add 1 to the value and return the new value. It's an atomic operation. The return value -1 is the initial value which is also the index.